langchain/chains/qa_with_sources/vector_db.py

```diff
@@ -15,11 +17,33 @@ class VectorDBQAWithSourcesChain(BaseQAWithSourcesChain, BaseModel):
     """Vector Database to connect to."""
     k: int = 4
     """Number of results to return from store"""
+    reduce_k_below_max_tokens: bool = False
+    """Reduce the number of results to return from store based on tokens limit"""
+    max_tokens_limit: int = 3375
+    """Restrict the docs to return from store based on tokens,
+    enforced only for StuffDocumentChain and if reduce_k_below_max_tokens is to true"""
     search_kwargs: Dict[str, Any] = Field(default_factory=dict)
     """Extra search args."""
 
+    def _page_content(self, doc: Document) -> str:
+        return doc.page_content
+
+    def _reduce_tokens_below_limit(self, docs: List[Document]) -> List[Document]:
```

Comment:
this looks good! one other way i was thinking of to implement this function could be the following:

```
def _reduce_tokens_below_limit(self, docs: List[Document]) -> List[Document]:

    num_docs = len(docs)
    
    if isinstance(self.combine_documents_chain, StuffDocumentsChain):
        tokens = [self.combine_documents_chain.llm_chain.llm.get_num_tokens(self._page_content(doc)) for doc in docs]

        while(sum(tokens[:num_docs])) > self.max_tokens_limit:
            num_docs -= 1

    return docs[:num_docs]
```
the advantage of this would be that we only have to call the `get_num_tokens` function once for each document, instead of multiple times. what do you think?

Reply:
Sorry if this is a n00b question - just familiarizing myself here 😄 - but would it make sense to avoid the quadratic run time? 

```python
def _reduce_tokens_below_limit(self, docs: List[Document]) -> List[Document]:

    num_docs = len(docs)
    
    if isinstance(self.combine_documents_chain, StuffDocumentsChain):
        tokens = [self.combine_documents_chain.llm_chain.llm.get_num_tokens(self._page_content(doc)) for doc in docs]

        token_count = sum(tokens[:num_docs])
        while token_count > self.max_tokens_limit:
            num_docs -= 1
            token_count -= tokens[num_docs]

    return docs[:num_docs]
```

Reply:
TL;DR: I'll make the change and go with what you recommend.

Though I wanted to share some thoughts on why I went with the recursive function design in the first place,

- This code was something I had implemented for my use case, so readability was crucial for me. I believe recursive code, when done right, is lovely :)
- The K value would always be in single or lower double digits, so I didn't have to worry about StackOverflow or other performance issues. 
- E.g., Doing a similarity search in a vector store can be slow, primarily if you use a remote index like Pinecone; I wanted to call them sparingly. That's why I do slicing on the docs list.
- On the other hand, I felt encoding would add a few milliseconds at best, and given the OpenAI response is usually in higher single-digit seconds, it would be a premature optimization.

Despite that, we must focus on the most performant code when implementing this functionality in a general-purpose library. Because we won't know if K will remain a small value tomorrow for newer LLM models, and most likely, we will have moved to other things by then, go back and fix this :)

So thank you so much for taking the time.

PS : @rowillia: Yours is better, but the overhead is small when the K value is small, especially since addition is a quick operation. Doing encoding again and again, which was my original sin is more expensive.

Reply:
Fixed.